File: android/edge-frameworks-core/src/main/kotlin/io/github/rajibsarwar/edgeframeworks/EdgeTool.kt
package io.github.rajibsarwar.edgeframeworks

/** A provider-neutral asynchronous operation exposed as a tool. */
interface EdgeTool {
    val definition: EdgeToolDefinition

    /**
     * Executes the tool with JSON-encoded arguments.
     *
     * Implementations decode and validate arguments and throw on failure.
     * Provider registration and model-driven dispatch belong to provider adapters.
     */
    suspend fun call(argumentsJson: String): EdgeToolResult
}

File: ios/Sources/EdgeFrameworks/Core/EdgeTool.swift
/// A provider-neutral asynchronous operation exposed as a tool.
public protocol EdgeTool: Sendable {
    var definition: EdgeToolDefinition { get }

    /// Executes the tool with JSON-encoded arguments.
    ///
    /// Implementations decode and validate arguments and throw on failure.
    /// Provider registration and model-driven dispatch belong to provider adapters.
    func call(argumentsJSON: String) async throws -> EdgeToolResult
}
